Create rule S7172 Named methods should be used to avoid confusion between testing an optional or an expected and testing the wrapped value CPP-5929 (#4545)
SonarSource/rspec: rules/S7172/cfamily/rule.adoc

> FIXME: add a description

Reviewer: Remove it, and maybe add one setence summary here for the ise.

// If you want to factorize the description uncomment the following line and create the file.
//include::../description.adoc[]

== Why is this an issue?

> `std::optional`, `boost::optional`, and `std::expected` are types that allow to represent a contained value of a certain type, and additional data that denote the absence of a value (`nullopt` for `optional`, and error values for `expected`). To check if the wrapper contains a value, it is possible to call the member function `has_value()`. It is equivalent to converting the wrapper to `bool`, which can be more concise, especially when used in a place that allows _contextual conversion to bool_ (for instance, the condition of an `if`). It means that the following syntax is valid:

Reviewer: ```suggestion
`std::optional`, `boost::optional`, and `std::expected` are types that allow to represent a contained value of a certain type, and additional data that denote the absence of a value (`nullopt` for `optional`, and error values for `expected`). To check if the wrapper contains a value, it is possible to call the member function `has_value()`. Alernatively conversion to  `bool` may be used, which can be more concise when used in a place that allows _contextual conversion to bool_ (for instance, the condition of an `if`). It means that the following syntax is valid:
```

[source,cpp]
----
std::optional<bool> flag;
if(flag) { ... }
// Equivalent to 
if(flag.has_value()) { ... }
----

When the wrapped type is also convertible to `bool`, using this concise syntax can be confusing. What is tested: The wrapper, or the wrapped value?

> This rule raises an issue when `std::optional`, `boost::optional`, or `std::expected` wrap a basic type, and the presence of the value is tested with the concise syntax. However, if in a single expression, the presence of the value is tested and the value is accessed, the risk of confusion is reduced, and no violation is raised.

Reviewer: I would put the first sentence also before "Why this is an issue", also I think reusing defintion of concise syntax here is a bit confusing, so I would just describe it there.
```suggestion
This rule raises an issue when `std::optional`, `boost::optional`, or `std::expected` wrap a basic type, and the `bool` conversion is use to test presence of the value. However, if in a single expression, the presence of the value is tested and the value is accessed, the risk of confusion is reduced, and no violation is raised.
```
I would put second sentence in exception seciton and add an example.

     // perform recursion...
  }
}
----

Since the rule will not trigger if the value is accessed in the same expression where its presence is tested, another possibility if the following:

[source,cpp]
----
void perform(Node node, optional<bool> recursive, Action action) {
  action(node);
  bool recurse = recursive ? *recursive : getDefaultRecurseMode();
  if (recurse) {
     // perform recursion...
  }
}
----

In this simple case, the use of both `recursive` and `++*recursive++` hints that recursive is more than a simple `bool`. However, it is possible to write the code with a higher level semantic:

[source,cpp]
----
void perform(Node node, optional<bool> recursive, Action action) {
  action(node);
  bool recurse = recursive.value_or(getDefaultRecurseMode());
  if (recurse) {
     // perform recursion...
  }
}
----
The downside of this version is that `getDefaultRecurseMode()` is called even if `recursive` contains a value. If `getDefaultRecurseMode()` is a complex function, it can be a performance issue. In this case, it is possible to use the monadic interface of `optional`, at the cost of a more complex syntax:

[source,cpp]
----
void perform(optional<bool> recursive) {

>   bool recurse = recursive.or_else([]() -> optional<bool> { return getDefaultRecurseMode();}).value();

Reviewer: Maybe write it:
`recursive.or_else(&getDefaultRecurseMode).value();

Reviewer: As discussed does not compile.

> 

Reviewer: I would add a line here:
> This rule raises an issue when `std::optional`, `boost::optional`, or `std::expected` wrap a basic type, and the conversion to `bool` is used to test presence of the value.
In SonarLint it is just displayed below the title and before the tabs, saying while this is an issue. So it is good place, to give tldr of the rule

Reviewer: See for example here: https://rules.sonarsource.com/cpp/RSPEC-7127/?search=array
